Hubverse ms vignette and associated documents (#15)
hubverse-org/hubverse: vignettes/hubverse_ms_vignette.Rmd

---
title: "Supplemental Code Demo"
subtitle: "Accessing Data from the US Flusight Hubverse Project"
author: 
- Consortium of Infectious Disease Modeling Hubs
- Melissa Kerr
- Rebecca Borchering
- Alvaro Castro Rivadeneira
- Lucie Contamin
- Sebastian Funk
- Harry Hochheiser
- Emily Howerton
- Anna Krystalli
- Li Shandross
- Nicholas G Reich
date: "`r Sys.Date()`"
output: 
  pdf_document:
    citation_package: natbib
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE
)
```

# Software configuration

> The following demonstration shows how to access and analyze data from a real hub using `r R.version$string` and several of the hubverse R packages.[@the_consortium_of_infectious_disease_modeling_hubs_hubverse_2025] The following packages are on CRAN and may be installed using the standard `install.packages()` function: `hubEnsembles`, `aws.s3`, `arrow`, `ggplot2` and `dplyr`.

Reviewer: I also had to update some of the packages to have the vignette knitting, so it might be a good idea to indicate which minimal version are necessary for some packages to replicate the vignette (either at the beginning or the end of the document)

> The following demonstration shows how to access and analyze data from a real hub using `r R.version$string` and several of the hubverse R packages.[@the_consortium_of_infectious_disease_modeling_hubs_hubverse_2025] The following packages are on CRAN and may be installed using the standard `install.packages()` function: `hubEnsembles`, `aws.s3`, `arrow`, `ggplot2` and `dplyr`.

Reviewer: I understand, I am debating that question too. It's not necessary to install them because they are dependencies and install from hubverse packages, but we are still using them. So maybe I will not include the code to install them but make a note of the package used in the vignette.

> The following demonstration shows how to access and analyze data from a real hub using `r R.version$string` and several of the hubverse R packages.[@the_consortium_of_infectious_disease_modeling_hubs_hubverse_2025] The following packages are on CRAN and may be installed using the standard `install.packages()` function: `hubEnsembles`, `aws.s3`, `arrow`, `ggplot2` and `dplyr`.

Author: I think this is addressed satisfactorily with @annakrystalli 's updates to the vignette.

The hubverse packages `hubData`, `hubVis`, and `hubEvals` must be installed through the R-universe package repository using the commands below.
```{r install, eval=FALSE}
install.packages(pkgs = c("hubData", "hubVis", "hubEvals"),
                 repos = c("https://hubverse-org.r-universe.dev",
                           "https://cloud.r-project.org"))
```

Once the required packages are installed, load them into your R session. (Session information including package versions are included at the end of this vignette.)

```{r libraries, message=FALSE}
library(hubData)
library(hubEnsembles)
library(hubVis)
library(hubEvals)
```

```{r, echo=FALSE}
ggplot2::theme_set(ggplot2::theme_bw())
```

# Load forecast data
CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub).[@noauthor_cdcepiflusight-forecast-hub_2025] Using hubverse tools, it also has been mirrored to the cloud, so a copy of the data lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of the repository, the data for this project can be directly accessed from the cloud using `hubData` tools.

> 

Reviewer: Should we add a little more context to the hub, saying what the goal is quickly and that the data mirrored on the bucket have all passed validation (using hubValidations)

```{r}
(hub_con <- hubData::s3_bucket("cdcepi-flusight-forecast-hub/") |>
   hubData::connect_hub(skip_checks = TRUE))
```

We then collect a subset of model output that we will interact with further in this demonstration. Specifically, we will extract zero through three week-ahead quantile predictions of weekly incident flu hospitalizations made for the state of Texas on three dates in late 2023 and early 2024. For simplicity, we will restrict our query to predictions made for five arbitrarily chosen models, although predictions for over 30 models are available on each of these dates.

```{r}
models <- c("FluSight-baseline",
            "CMU-TimeSeries",
            "UMass-flusion",
            "cfa-flumech",
            "UGA_flucast-INFLAenza")
reference_dates <- c("2023-12-16", "2024-01-13", "2024-02-10")

flusight_data <- hub_con |>
  dplyr::filter(target == "wk inc flu hosp",
                reference_date %in% reference_dates,
                model_id %in% models,
                output_type == "quantile",
                location == "48", ## FIPS code for Texas
                horizon > -1) |>
  hubData::collect_hub()
```

```{r, echo=FALSE}

> knitr::kable(flusight_data[116:120,
>                            c("model_id", "reference_date", "location",
>                              "output_type", "output_type_id", "value")],
>              digits = 2,
>              caption = "Five rows of the `flusight_data` object collected above.
>              The \"target\", \"horizon\", and \"target_end_date\" columns have
>              been excluded for readability. The \"output_type_id\" and \"value\"
>              columns show the different predicted quantiles for the number of
>              hospital admissions due to influenza made by the CMU-TimeSeries
>              model for location 48 (Texas). The predictions were made during the
>              week of December 16, 2023 (reference date) and are making a
>              prediction for the horizon of 1 week ahead, for the week ending
>              December 23.")

Reviewer: ```suggestion
knitr::kable(flusight_data[116:120,
                           c("model_id", "reference_date", "horizon",
                             "output_type", "output_type_id", "value")],
             digits = 2,
             caption = "Five rows of the `flusight_data` object collected above.
             The \"target\", \"location\", and \"target_end_date\" columns have
             been excluded for readability. The \"output_type_id\" and \"value\"
             columns show the different predicted quantiles for the number of
             hospital admissions due to influenza made by the CMU-TimeSeries
             model for location 48 (Texas). The predictions were made during the
             week of December 16, 2023 (reference date) and are making a
             prediction for the horizon of 1 week ahead, for the week ending
             December 23.")
```
As the previous filter focus on only Texas, I thought maybe it might make it easier for the user to show horizon then location here. As horizon will be expected to change but not location.

```

# Loading target data

> We can use the `aws.s3` package to read the contents of the most up-to-date target data file stored in the cloud.

Reviewer: Should it be updated to match the latest requirement/standard from hubverse? (to avoid rename, etc.)

> We can use the `aws.s3` package to read the contents of the most up-to-date target data file stored in the cloud.

Reviewer: ```suggestion
We can use the `aws.s3` package to read the contents of the most up-to-date target data (also called "observed data") file stored in the cloud.
```
I suggest we add "observed data" because it's used in the next "visualization" part of the vignette.

             The observation column shows the number of hospital admissions
             due to influenza reported in Texas (location 48) on the week
             ending on each date.")
```

\clearpage 

## Visualize forecasts

We can use the hubVis package to create a single graphic that overlays the forecasts from the five models and three forecast dates with the observed data.

```{r, fig.cap = "Forecasts of hospital admissions in Texas due to influenza in the 2023-2024 respiratory virus season. Predictions from five models are shown along with observed data (line with dots for specific observations). Each model is shown in a different color, with the 80% prediction intervals shown as a shaded region.", fig.width = 9, fig.height = 6}  
flusight_data |>
  hubVis::plot_step_ahead_model_output(
    target_data,
    x_col_name = "target_end_date",
    use_median_as_point = TRUE,
    group = "reference_date",
    interactive = FALSE,
    fill_transparency = .1,
    intervals = 0.8
  ) +
  ggplot2::theme(legend.position = "bottom") +
  ggplot2::ylab("hospital admissions due to influenza")
```

\clearpage 

# Build ensemble and visualize

> Using the model output data collected above, we can pass these data to the `hubEnsembles::simple_ensemble()` function to build an ensemble that takes the mean of predicted values from each model at each quantile level. This new forecast output is then added to the existing forecasts and visualized all together. The hubEnsembles package supports a range of aggregation approaches including a linear opinion pool (probability density averaging) or using the mean or median of quantiles.[@shandross_multi-model_2025] Model weights can be specified for the various approaches, allowing for some models to influence the output more strongly.

Reviewer: ```suggestion
Using the model output data collected above, we can pass these data to the `hubEnsembles::simple_ensemble()` function to build an ensemble that takes the mean of predicted values from each model at each quantile level. This new forecast output is then added to the existing forecasts and visualized all together. The `hubEnsembles` package supports a range of aggregation approaches including a linear opinion pool (probability density averaging) or using the mean or median of quantiles.[@shandross_multi-model_2025] Model weights can be specified for the various approaches, allowing for some models to influence the output more strongly.
```
(following @eahowerton suggestion)

> ```{r, fig.cap = "Forecasts of hospital admissions in Texas due to influenza in the 2023-2024 respiratory virus season. Predictions from five submitted models and an ensemble are shown along with observed data. The observed data are shown as a line with dots for specific weekly observations. Each individual model is shown in orange and the mean ensemble is shown in black, with the 80% prediction intervals shown as shaded regions."}

Reviewer: ```suggestion
```{r, fig.cap = "Forecasts of hospital admissions in Texas due to influenza in the 2023-2024 respiratory virus season. Predictions from five submitted models and an ensemble are shown along with observed data. The observed data are shown as a line with dots for specific weekly observations. Each individual model is shown in gray and the mean ensemble is shown in orange, with the 80% prediction intervals shown as shaded regions."}
```
(following @eahowerton  suggestion)

ensemble_forecast <- hubEnsembles::simple_ensemble(flusight_data)

all_forecasts <- flusight_data |>
  dplyr::bind_rows(ensemble_forecast)

all_forecasts |>
  hubVis::plot_step_ahead_model_output(target_data,
                                       x_col_name = "target_end_date",
                                       use_median_as_point = TRUE,
                                       group = "reference_date",
                                       interactive = FALSE,
                                       fill_transparency = .2,

>                                        intervals = 0.90,

Reviewer: ```suggestion
                                       intervals = 0.80,
```

                                       one_color = "darkgray",
                                       ens_name = "hub-ensemble",
                                       ens_color = "orange",
                                       pal_color = NULL) +
  ggplot2::theme(legend.position = "bottom") +
  ggplot2::ylab("hospital admissions due to influenza")
```

\clearpage

# Evaluate forecasts

> The package hubEvals can be used to evaluate probabilistic predictions against observed data. The hubEvals package creates an interface for hubverse-style model output and target data to the scoringutils package, which is an engine for computing a number of common scores for probabilistic predictions.[@bosse-scoringutils-2024] 

Reviewer: ```suggestion
The package `hubEvals` can be used to evaluate probabilistic predictions against observed data. The `hubEvals` package creates an interface for hubverse-style model output and target data to the `scoringutils` package, which is an engine for computing a number of common scores for probabilistic predictions.[@bosse-scoringutils-2024] 
```

> The code below starts by making oracle output data, which is a dataset formatted like hubverse-style model output data but with a point prediction representing the eventual observation, as if observed by an oracle model. The FluSight hub stores target data in a time-series format but not in an oracle output format. The oracle output data format is convenient to use for scoring, as it can be joined with model output data and then the observation and the prediction reside in a single row of stored data. In this example, we compute and show results for the weighted interval score (WIS) with its associated decomposition into overprediction, underprediction, and dispersion penalties, the 50% and 90% prediction interval coverage rates, and the absolute error computed on the median prediction.[@bracher_evaluating_2021]

Reviewer 2: I think this should be "average weighted interval score"

> The code below starts by making oracle output data, which is a dataset formatted like hubverse-style model output data but with a point prediction representing the eventual observation, as if observed by an oracle model. The FluSight hub stores target data in a time-series format but not in an oracle output format. The oracle output data format is convenient to use for scoring, as it can be joined with model output data and then the observation and the prediction reside in a single row of stored data. In this example, we compute and show results for the weighted interval score (WIS) with its associated decomposition into overprediction, underprediction, and dispersion penalties, the 50% and 90% prediction interval coverage rates, and the absolute error computed on the median prediction.[@bracher_evaluating_2021]

Reviewer: Should we add a little bit more, by saying that you are using the `hubEvals::score_model_out()` function that scores model outputs with a single output_type against observed data, and more information is available in the help page of the function and the associated package?

The hubverse packages `hubData`, `hubVis`, `hubEnsembles` and `hubEvals` are best installed through the R-universe package repository to ensure you are working with the latest released versions. Note that these packages have some dependencies that are needed for the operations in this vignette, so specifying `dependencies = TRUE` is recommended. The packages can be installed from the R console using the following command:

```{r install, eval=FALSE}
install.packages(
  pkgs = c("hubData", "hubVis", "hubEnsembles", "hubEvals"),
  repos = c(
    "https://hubverse-org.r-universe.dev",
    "https://cloud.r-project.org"
  ),
  dependencies = TRUE
)
```

Supporting packages `ggplot2` and `dplyr` can be installed from CRAN.

```{r install-CRAN, eval=FALSE}
install.packages(
  pkgs = c("dplyr", "ggplot2")
)
```

Once the required packages are installed, load the supporting packages into your R session. (Session information on all the packages used including package versions are included at the end of this vignette.)

```{r libraries, message=FALSE}
# Data munging
library(dplyr)
# Plotting
library(ggplot2)
```

# Load forecast data

> CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub)[@noauthor_cdcepiflusight-forecast-hub_2025]. Using hubverse tools, the data for this hub has been validated and mirrored to the cloud, so a copy of the model output for this hub hub lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of the repository, the data for this project can be directly accessed from the cloud using `hubData` tools.

Reviewer 3: ```suggestion
CDC's FluSight Forecast Hub exists as a [GitHub repository](https://github.com/cdcepi/FluSight-forecast-hub)[@noauthor_cdcepiflusight-forecast-hub_2025]. Using hubverse tools, the data for this hub has been validated and mirrored to the cloud, so a copy of the model output for this hub lives at a publicly-accessible AWS S3 bucket. This means that instead of needing to download a local copy of the repository, the data for this project can be directly accessed from the cloud using `hubData` tools. The data is organised in a format we can open as an arrow dataset through R package `arrow` directly .
```

```{r}
hub_path <- hubData::s3_bucket("cdcepi-flusight-forecast-hub/")
hub_con <- hubData::connect_hub(hub_path, skip_checks = TRUE)
hub_con
```

The hub connection object printed above shows information about the hub, including the number and format of files loaded, the schema of the data, and column names and data types of model-output data. Note that the `skip_checks = TRUE` argument is used to skip the validation checks that are normally performed when connecting to a hub. In practice, since these data are all validated upon submission to the hub and since the checks impact performance, we omit the checks in this example.

We then collect a subset of model outputs that we will interact with further in this demonstration. Specifically, we will extract zero through three week-ahead quantile predictions of weekly incident flu hospitalizations made for the state of Texas on three dates in late 2024 and early 2025. For simplicity, we will restrict our query to predictions made for five arbitrarily chosen models, although predictions for over 30 models are available on each of these dates. 

> While the files are stored in the hub separately by model and reference date, the `hubData::collect_hub()` function, through functionality from `arrow`, allows us to act on all of the data collectively, as if if were part of a single `data.frame` object.

Reviewer 3: ```suggestion
Although the data are stored in the hub across multiple files, organised by `model_id` and `round_id`, `hubData::connect_hub()` wraps functionality from `arrow` to open a connection to the data as an Arrow Dataset. This allows us to act on all of the data collectively, as if it were part of a single `data.frame` object.  Arrow datasets, including those stored on S3, behave as lazy data frames—they do not load data into memory until explicitly queried or collected. The `arrow` package also provides a [dplyr](https://dplyr.tidyverse.org/) interface to Arrow Datasets which allows us to query and manipulate the data.
```

```{r}
models <- c(
  "FluSight-baseline",
  "CMU-TimeSeries",
  "UMass-flusion",
  "cfa-flumech",
  "UGA_flucast-INFLAenza"
)

reference_dates <- c("2024-12-14", "2025-01-11", "2025-02-08")

flusight_data <- hub_con |>
  filter(
    target == "wk inc flu hosp",
    reference_date %in% reference_dates,
    model_id %in% models,
    output_type == "quantile",
    location == "48", ## FIPS code for Texas
    horizon > -1
  ) |>
  hubData::collect_hub()
```

> 

Reviewer 3: ```suggestion
The `collect_hub` `hubData` function wraps the `dplyr::collect()` function but also coerces the output to a `model_out_tbl` class object where possible.
```

Hubs generally contain two types of target data: time-series data and oracle output data. Time-series data is stored in a long format, with one row per observation, a single observed value per observation unit and is the format used for visualization of observed target data. Oracle output data on the other hand is often derived from time-series data but more closely matches the model output data format and is used for the evaluation of model outputs. The FluSight hub stores target data in both formats, but we will first load the time-series format to visualize the performance of hub models against observed data. For more information on target data formats, see the [hubverse documentation](https://docs.hubverse.io/en/latest/user-guide/target-data.html).

Accessing target data follows a similar pattern to accessing model output data. We start by connecting to the target data type of interest in the hub. To access the time-series format we use `hubData::connect_target_timeseries()` function.

```{r}
ts_data_con <- hubData::connect_target_timeseries(hub_path)
```

> Next, we filter the target data to the location and time period of interest. In this case, we are interested in the state of Texas (FIPS code 48) and the 2024-2025 respiratory virus season, which runs from September 23, 2024 to May 1, 2024. We also filter to the most recent `as_of` date to ensure we are using the latest available data. The `as_of` date is a metadata field that indicates when the target data was last updated.

Reviewer 3: ```suggestion
Next, we filter the target data to the location and time period of interest. In this case, we are interested in the state of Texas (FIPS code 48) and the 2024-2025 respiratory virus season, which runs from September 23, 2024 to May 1, 2025. We also filter to the most recent `as_of` date to ensure we are using the latest available data. The `as_of` date is a metadata field that indicates when the target data was last updated.
```

The package `hubEvals` can be used to evaluate probabilistic predictions against observed data. The `hubEvals` package creates an interface for hubverse-style model output and oracle output target data to the `scoringutils` package, which is an engine for computing a number of common scores for probabilistic predictions[@bosse-scoringutils-2024].

As mentioned earlier, to evaluate model outputs we use oracle output data, a form of target data which matches and can be merged with hubverse-style model output data but which contains a point prediction representing the eventual observation, as if observed by an oracle model. 

The FluSight hub also stores oracle output data in the cloud, and we can access it using the `hubData::connect_target_oracle_output()` function. This function connects to the oracle output data in the hub and returns a connection object that can again be used to query the data.

```{r}
oo_data_con <- hubData::connect_target_oracle_output(hub_path)
```

> 

Reviewer 3: Note that the oracle-output contains a `horizon` column, which [as discussed](https://github.com/reichlab/decisions/pull/31#discussion_r2175444614), probably shouldn't be there?

I think we should resolve this or at least open an issue about resolving it soon in. I don't feel we should be showcasing wrongly formatted hubs in the vignette as it can cause unnecessary confusion.
